distributor: do inflight checks before reading request (for httpgrpc requests)

pstibrany opened the pull request:
#### What this PR does

This PR implements inflight checks in Distributor by using gRPC method limiter introduced in https://github.com/grafana/dskit/pull/377. This PR relies on unmerged dskit PR https://github.com/grafana/dskit/pull/392 (waiting for new gRPC-go release).

This PR also relies on clients passing the correct metadata headers for httpgrpc requests (using methods introduced in https://github.com/grafana/dskit/pull/391).

To be rebased on top of https://github.com/grafana/mimir/pull/6301, after it's merged.

#### Checklist

- [x] Tests updated
- [na] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on pkg/mimir/grpc_push_check.go:
```diff
@@ -3,44 +3,124 @@
+		httpUrl := getSingleMetadata(md, httpgrpc.MetadataURL)
+		msgSize := getMessageSize(md, grpcutil.MetadataMessageSize)
+
+		if httpMethod == http.MethodPost && (httpUrl == api.PrometheusPushEndpoint || httpUrl == api.OTLPPushEndpoint) {
```
I'm not sure checking the URL for equality will work. Reason is that the endpoint is exposed under a configurable prefix (and I think the prefix is still the URL received here). What if we check if the URL ends with the requested endpoint? That's [what we do in the query-frontend](https://github.com/grafana/mimir/blob/dec27843609f58ecc0072e9cc222e126aee255e0/pkg/frontend/querymiddleware/roundtrip.go#L368-L382).

pstibrany replied:
Good point, done in https://github.com/grafana/mimir/pull/6300/commits/5cd002d94efdab78e581967f0d28b52ab344b61a.

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -1006,36 +1010,151 @@ func (d *Distributor) metricsMiddleware(next push.Func) push.Func {
+	inflight := d.inflightPushRequests.Inc()
+	defer func() {
+		if cleanupInDefer {
+			d.inflightPushRequests.Dec()
```
I'm wondering if to keep the code more future proof, we should decrease `d.inflightPushRequestsBytes` here. My take is that in in the future we'll add more checks after `d.checkRequestSize()` then we'll miss subtracting it. The check here could be done on `rs.requestSize`.

pracucci replied:
Actually we could just change the defer function into a call to `cleanupAfterPushFinished()`.

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -187,7 +189,8 @@ type Config struct {
 
-	WriteRequestsBufferPoolingEnabled bool `yaml:"write_requests_buffer_pooling_enabled" category:"experimental"`
+	WriteRequestsBufferPoolingEnabled           bool `yaml:"write_requests_buffer_pooling_enabled" category:"experimental"`
+	LimitInflightRequestsUsingGrpcMethodLimiter bool `yaml:"limit_inflight_requests_using_grpc_method_limiter" category:"experimental"`
```
Looks like it's not used. I guess it's just one of those piece not implemented yet.

pstibrany replied:
It's used from here: https://github.com/grafana/mimir/blob/bc26fa26a24f33e00d7e546032904294569dfa13/pkg/mimir/modules.go#L291

pstibrany commented:
I think this is now ready for review. Compared to previous version of the PR, we now check size of `httpgrpc.Request` and `mimirpb.WriteRequest` separately, and they *both* contribute to inflight request size, because they are both in memory while request is being handled.

Extension of this (for future PR) could be to check for OTLP Request size too in OTLP Push handler.

pracucci commented on CHANGELOG.md:
```diff
@@ -49,6 +49,7 @@
 * [ENHANCEMENT] Querier: reduce memory consumed for queries that hit store-gateways. #6348
 * [ENHANCEMENT] Ruler: include corresponding trace ID with log messages associated with rule evaluation. #6379
 * [ENHANCEMENT] Querier: clarify log messages and span events emitted while querying ingesters, and include both ingester name and address when relevant. #6381
+* [ENHANCEMENT] Ingester, Distributor: added experimental support for rejecting push requests received via gRPC before reading them into memory, if ingester or distributor is unable to accept the request. This is activated by using `-ingester.limit-inflight-requests-using-grpc-method-limiter` for ingester, and `-distributor.limit-inflight-requests-using-grpc-method-limiter` for distributor. #5976 ...
```
Mention the experimental flags in `docs/sources/mimir/configure/about-versioning.md` too, please.

pstibrany replied:
https://github.com/grafana/mimir/pull/6300/commits/304de04c7169aed6b1e4330c07e7f06077c853c7

pracucci commented on pkg/mimir/grpc_push_check.go:
```diff
@@ -48,13 +68,61 @@ func (g *grpcInflightMethodLimiter) RPCCallStarting(ctx context.Context, methodN
+		// Let's check httpgrpc metadata, if present.
+		httpMethod := getSingleMetadata(md, httpgrpc.MetadataMethod)
+		httpUrl := getSingleMetadata(md, httpgrpc.MetadataURL)
+		msgSize := getMessageSize(md, grpcutil.MetadataMessageSize)
```
[nit] We could move this inside the `if` branch.

pracucci commented on pkg/mimir/grpc_push_check_test.go:
```diff
@@ -50,21 +66,133 @@ func TestGrpcInflightMethodLimiter(t *testing.T) {
+
+		ctx, err = l.RPCCallStarting(context.Background(), httpgrpcHandleMethod, metadata.New(map[string]string{
+			httpgrpc.MetadataMethod:      "POST",
+			httpgrpc.MetadataURL:         api.PrometheusPushEndpoint,
```
[nit] Would be nice to add a prefix to ensure it matches anyway.

pstibrany replied:
https://github.com/grafana/mimir/pull/6300/commits/e65144f2d4e0550167f294b66a61e740c261d230

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -1005,36 +1007,164 @@ func (d *Distributor) metricsMiddleware(next PushFunc) PushFunc {
+	il := d.getInstanceLimits()
+	if il.MaxInflightPushRequests > 0 && inflight > int64(il.MaxInflightPushRequests) {
+		d.rejectedRequests.WithLabelValues(reasonDistributorMaxInflightPushRequests).Inc()
+		return ctx, nil, errMaxInflightRequestsReached
```
Can you confirm the reason why we removed the wrapping to `util_log.DoNotLogError{}` is because all errors returned by this function will not be logged anyway? If so, I would add a comment to clarify it.

pstibrany replied:
Added comment. Wrapping is still done in `limitsMiddleware`.

pracucci commented on pkg/distributor/distributor.go:
```diff
@@ -1005,36 +1007,164 @@ func (d *Distributor) metricsMiddleware(next PushFunc) PushFunc {
+
+func (d *Distributor) checkWriteRequestSize(rs *requestState, writeRequestSize int64) error {
+	// If writeRequestSize was already checked, don't check it again.
+	if rs.writeRequestSize > 0 {
```
Question: I'm good with this check, but could it ever happen that we call `checkWriteRequestSize()` twice for the same request? If yes, in which case? I can't see it.

pstibrany replied:
I don't see how that would happen. It's *only* called from `limitsMiddleware`, and not from grpc method check. (We only call `checkHttpgrpcRequestSize` from there).

pracucci reviewed (approved):
Good job Peter! I can't see any issue. I left a couple of nits and questions (don't block on me re-reviewing it).

pstibrany commented:
> Good job Peter! I can't see any issue. I left a couple of nits and questions (don't block on me re-reviewing it).

Thanks for review. I've addressed your feedback, setting auto-merge. 🎉
